### PR TITLE
Adding screen with pending retries

### PR DESCRIPTION
**Pull request by WojcikMike**
Relates to: Particular/V6Launch#77

### CODE REVIEW

File: src/ServicePulse.Host/app/js/directives/ui.particular.configurationTabs.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
+﻿
+; (function (window, angular, $, undefined) {
+    'use strict';
+
+
+    function controller($scope, $interval, $location, redirectService, notifyService, sharedDataService) {
+        var notifier = notifyService();
+
+        $scope.isActive = function (viewLocation) {
+            var active = (viewLocation === $location.path());
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/directives/ui.particular.configurationTabs.js]: Why the local variable?

File: src/ServicePulse.Host/app/js/directives/ui.particular.configurationTabs.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        var stats = sharedDataService.getstats();
+
+        redirectService.getTotalRedirects().then(function(response) {
+            $scope.counters.redirects = response || 0;
+        });
+
+        $scope.counters = {
+            endpoints: stats.number_of_endpoints,
+            redirects: 0
+        }
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/directives/ui.particular.configurationTabs.js]: Missing semicolon

File: src/ServicePulse.Host/app/js/directives/ui.particular.configurationTabs.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+        var redirectPromise = $interval(function () {
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/directives/ui.particular.configurationTabs.js]: This isn't a promise, it's an interval timer. Perhaps rename to `redirectUpdatedTimer`? Or something more meaningful?

File: src/ServicePulse.Host/app/js/directives/ui.particular.configurationTabs.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    function directive() {
+        return {
+            scope: {},
+            restrict: 'E',
+            replace: true,
+            templateUrl: 'js/directives/ui.particular.configurationTabs.tpl.html',
+            controller: controller,
+            link: function (scope, element) { }
+        };
+    };
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/directives/ui.particular.configurationTabs.js]: Semicolon not needed here (function definition)

File: src/ServicePulse.Host/app/js/directives/ui.particular.failedMessageTabs.js [hunk 5]
```diff
@@ -18,8 +18,9 @@
         $scope.counters = {
             group: stats.number_of_exception_groups,
             message: stats.number_of_failed_messages,
-            archived: stats.number_of_archived_messages
-        }
+            archived: stats.number_of_archived_messages,
+            pendingRetries: stats.number_of_pending_retries
+    }
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/directives/ui.particular.failedMessageTabs.js]: Indentation here looks off, should be 4 spaces in perhaps?

File: src/ServicePulse.Host/app/js/directives/ui.particular.failedMessageTabs.js [hunk 6]
```diff
@@ -38,6 +39,12 @@
             });
         }, 10000);
 
+        var pendingRetriesPromise = $interval(function () {
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/directives/ui.particular.failedMessageTabs.js]: Again, this isn't a promise but an interval timer.

File: src/ServicePulse.Host/app/js/services/services.service-control.js [hunk 7]
```diff
@@ -161,6 +194,22 @@
...
         }
 
+        function markAsResolvedMessages(selectedMessages) {
+            var url = uri.join(scConfig.service_control_url, 'errors', 'resolve');
+
+            $http({
+                url: url,
+                data: selectedMessages,
+                method: 'PATCH'
+            })
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/services/services.service-control.js]: Indentation looks weird here again

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+
+                vm.pendingRetryMessages = vm.pendingRetryMessages.concat(exgroups);
+                vm.allMessagesLoaded = (vm.pendingRetryMessages.length >= vm.total);
+                vm.page++;
+            }
+            vm.loadingData = false;
+        };
+
+        var init = function() {
+            vm.pendingRetryMessages = [];
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: Duplicating some of the init code above (line 19 -- 29). Perhaps combine?

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+            return serviceControlService.getQueueNames(searchPhrase).then(function(results) {
+                return results.map(function(item) {
+                    return item.physical_address;
+                });
+            });
+        };
+
+        vm.clipComplete = function (messageId) {
+            toastService.showInfo(messageId + ' copied to clipboard');
+        }
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: Missing semicolon

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+
+        vm.togglePanel = function (message, panelnum) {
+            if (message.messageBody === undefined) {
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: Probably better to stick to using the `angular.isDefined` convetion

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+                serviceControlService.getMessageBody(message.message_id).then(function (msg) {
+                    message.messageBody = msg.data;
+                }, function () {
+                    message.bodyUnavailable = "message body unavailable";
+                });
+            }
+
+            if (message.messageHeaders === undefined) {
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: angular.isDefined

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+                }, function () {
+                    message.headersUnavailable = "message headers unavailable";
+                });
+            }
+            message.panel = panelnum;
+            return false;
+        };
+
+        vm.toggleSelectAll = function() {
+            for (var i = 0; i < vm.pendingRetryMessages.length; i++) {
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]:
````
vm.pendingRetryMessages.filter(function (item) {
    return item.selected !== vm.allSelected;
}).forEach(function(item) {
    toggleSelect(item);
});
```

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+            toggleSelect(row);
+
+            vm.allSelected = vm.selectedIds.length === vm.pendingRetryMessages.length;
+        };
+
+        vm.retrySelected = function () {
+            serviceControlService.retryFailedMessages(vm.selectedIds);
+            vm.selectedIds = [];
+
+            for (var i = 0; i < vm.pendingRetryMessages.length; i++) {
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]:
```
vm.pendingRetryMessages.filter(function (item) {
    return item.selected;
}).forEach(function (item) {
    item.selected = false;
    item.retried = true;
});
```

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+                }
+            }
+        };
+
+        vm.markAsResolvedSelected = function () {
+            serviceControlService.markAsResolvedMessages(vm.selectedIds);
+            vm.selectedIds = [];
+
+            for (var i = 0; i < vm.pendingRetryMessages.length; i++) {
+                if (vm.pendingRetryMessages[i].selected) {
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: filter and forEach

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+                    vm.pendingRetryMessages[i].resolved = true;
+                }
+            }
+        };
+
+        vm.archiveSelected = function () {
+            serviceControlService.archiveFailedMessages(vm.selectedIds);
+            vm.selectedIds = [];
+
+            for (var i = 0; i < vm.pendingRetryMessages.length; i++) {
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: filter and forEach.

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+            }
+        };
+
+        vm.searchPhraseChanged = function() {
+            if (!vm.searchPhrase) {
+                vm.pendingRetryMessages = [];
+                vm.page = 1;
+                vm.loadMoreResults();
+            }
+        }
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: Missing semicolon

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+
+        vm.onSelect = function () {
+            vm.pendingRetryMessages = [];
+            vm.page = 1;
+            vm.loadMoreResults();
+        }
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: Missing semicolon

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+            if (vm.allMessagesLoaded || vm.loadingData) {
+                return;
+            }
+
+            vm.loadingData = true;
+            
+            serviceControlService.getPendingRetryMessages(vm.searchPhrase || '', vm.sort, vm.page, vm.direction).then(function (response) {
+                processLoadedMessages(response.data);
+            });
+        }
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: Missing semicolon

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+            $window.open("si://" + dnsName + "?search=" + messageId);
+        };
+
+        var selectGroupInternal = function (sort, direction, changeToMessagesTab) {
+            vm.sort = sort;
+            vm.direction = direction;
+            setSortButtonText(sort, direction);
+
+            if ($scope.loadingData) {
+                return;
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: Do we want the sort button text to be set if the code breaks out without actually loading the results? i.e. in the case where `$scope.loadingData` is true

File: src/ServicePulse.Host/app/js/views/redirect/controller.js [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+                            saveButtonText: saveButtonText,
+                            success: success,
+                            failure: failure
+                        };
+                    }
+                }
+            }).result.then(function () {
+                refreshData();
+            });
+        };
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/redirect/controller.js]: Don't need this semicolon

File: src/ServicePulse.Host/app/js/views/redirect/edit/controller.js [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+                    if ((response.status === '409' || response.status === 409) && response.statusText === 'Duplicate') {
+                        toastService.showError('Failed to create a redirect, can not create more than one redirect for queue: ' + $scope.from_physical_address);
+                    } else if ((response.status === '409' || response.status === 409) && response.statusText === 'Dependents') {
+                        toastService.showError('Failed to update a redirect, can not create redirect to a queue that already has a redirect or is a target of a redirect.');
+                    } else{
+                        toastService.showError(response.message);
+                    }
+                });
+            }
+        }
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/redirect/edit/controller.js]: Missing semicolon

File: src/ServicePulse.Host/app/js/views/redirect/route.js [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+    function routeProvider($routeProvider) {
+        $routeProvider.when('/redirects', {
+            data: {
+                pageTitle: 'Retry Redirects'
+            },
+            templateUrl: 'js/views/redirect/view.html',
+            controller: 'redirectController',
+            controllerAs: 'vm'
+        });
+    };
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/redirect/route.js]: Don't need this semicolon

File: src/ServicePulse.Host/app/js/views/configuration/configuration.controller.js [hunk 23]
```diff
@@ -12,14 +12,18 @@
         $scope,
         $window,
         sharedDataService,
-        configurationService) {
+        configurationService,
+        notifyService) {
 
+        var notifier = notifyService();
```
Comment (WojcikMike) [inline src/ServicePulse.Host/app/js/views/configuration/configuration.controller.js]: I agree, we should clean it up. However we should do it in a separate PR

File: src/ServicePulse.Host/app/js/directives/ui.particular.redirectLink.js [hunk 24]
```diff
@@ -0,0 +1,37 @@
+﻿
+; (function (window, angular, undefined) {
+    'use strict';
+
+
+    function controller($scope, redirectService) {
+        
+        $scope.editRedirect = function () {
+            redirectService.displayEditRedirectModal($scope.redirect);
+        }
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/directives/ui.particular.redirectLink.js]: Semicolon

File: src/ServicePulse.Host/app/js/directives/ui.particular.redirectLink.js [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+
+        $scope.createRedirect = function () {
+            redirectService.displayCreateRedirectModal();
+        }
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/directives/ui.particular.redirectLink.js]: Semicolon

File: src/ServicePulse.Host/app/js/directives/ui.particular.redirectLink.js [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+    function directive() {
+        return {
+            scope: { redirect: '='},
+            restrict: 'E',
+            replace: true,
+            templateUrl: 'js/directives/ui.particular.redirectLink.tpl.html',
+            controller: controller,
+            link: function (scope, element) { }
+        };
+    };
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/directives/ui.particular.redirectLink.js]: No semicolon :)

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js (hunk 8 above)
Comment (WojcikMike) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: Yeah I left the code in the init. if you have better idea let me know

**Comment by WojcikMike** [comment]
@WilliamBZA I applied all your comments. I left them for now in separate commits for ease of reviewing. I will combine them before merging

File: src/ServicePulse.Host/app/lib/bootstrap-3.3.5/css/bootstrap.css [hunk 27]
```diff
@@ -1583,8 +1583,8 @@ pre code {
   overflow-y: scroll;
 }
 .container {
-  padding-right: 15px;
-  padding-left: 15px;
+  padding-right: 0;
+  padding-left: 0;
```
Comment (WojcikMike) [inline src/ServicePulse.Host/app/lib/bootstrap-3.3.5/css/bootstrap.css]: @sergioc is there any way for us to not change bootstrap.css? For example to override this value in our particular.css file? If we upgrade bootstrap those changes will be gone.

Comment (sergioc) [inline src/ServicePulse.Host/app/lib/bootstrap-3.3.5/css/bootstrap.css]: We could add a custom class with those properties, but the same problem would remain. However, I don't see bootstrap 3.x breaking those properties. I also don't see a reason to upgrade bootstrap.

Comment (WojcikMike) [inline src/ServicePulse.Host/app/lib/bootstrap-3.3.5/css/bootstrap.css]: We can override those settings i particular.css (set paddings to 0). I verified it and it works. We could update bootstrap when we would want to benefit of any of the bug fixes or new features that they release. For 3.3.6 its: 
https://github.com/twbs/bootstrap/releases/tag/v3.3.6

What is more, we can decide to start using minified css which we would have to update with those values. I find it risky to modify 3rd party libraries. I am happy to override the styles if you don't mind

File: src/ServicePulse.Host/app/lib/bootstrap-3.3.5/css/bootstrap.css [hunk 28]
```diff
@@ -1583,8 +1583,8 @@ pre code {
   overflow-y: scroll;
 }
 .container {
-  padding-right: 0;
-  padding-left: 0;
+  padding-right: 15px;
+  padding-left: 15px;
```
Comment (sergioc) [inline src/ServicePulse.Host/app/lib/bootstrap-3.3.5/css/bootstrap.css]: @WojcikMike what's the need for this change?

File: src/ServicePulse.Host/app/lib/bootstrap-3.3.5/css/bootstrap.css [hunk 29]
```diff
@@ -1609,7 +1609,10 @@ pre code {
   margin-right: auto;
   margin-left: auto;
 }
-
+.row {
+  margin-right: -15px;
+  margin-left: -15px;
+}
```
Comment (sergioc) [inline src/ServicePulse.Host/app/lib/bootstrap-3.3.5/css/bootstrap.css]: And this?

File: src/ServicePulse.Host/app/lib/bootstrap-3.3.5/css/bootstrap.css (hunk 28 above)
Comment (WojcikMike) [inline src/ServicePulse.Host/app/lib/bootstrap-3.3.5/css/bootstrap.css]: @sergioc this is the reason: https://github.com/Particular/ServicePulse/pull/364#discussion_r74415589

I would be happy to discuss it on slack

Comment (sergioc) [inline src/ServicePulse.Host/app/lib/bootstrap-3.3.5/css/bootstrap.css]: Ah! I remember the discussion, but missed that last comment.

I checked the changes and it's all good, so 👍

**Comment by WojcikMike** [comment]
@WilliamBZA we are close to being done, can you please have a look at the code?

File: src/ServicePulse.Host/app/js/directives/ui.particular.failedMessageTabs.js [hunk 30]
```diff
@@ -1,15 +1,15 @@
 ﻿
-; (function (window, angular, $, undefined) {
+;
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/directives/ui.particular.failedMessageTabs.js]: Leading semicolon

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 31]
```diff
@@ -0,0 +1,375 @@
...
+        scConfig,
+        toastService,
+        sharedDataService,
+        notifyService,
+        serviceControlService,
+        endpointsService, 
+        redirectService,
+        pendingRetryService) {
+
+        var vm = this;
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: I think this should be indented one more time? It seems to be at the same indentation level as the `controller`

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 32]
```diff
@@ -0,0 +1,375 @@
...
+        vm.sortDirection = 'asc';
+        vm.allMessagesLoaded = false;
+        vm.loadingData = false;
+
+        vm.timeGroup = {
+            amount: undefined,
+            unit: undefined,
+            buttonText: 'All Pending Retries',
+            selected: function () {
+                return $moment.duration(vm.timeGroup.amount, vm.timeGroup.unit);;
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: Double semicolo

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 33]
```diff
@@ -0,0 +1,375 @@
...
+        vm.sortButtonText = '';
+        vm.sortDirection = 'asc';
+        vm.allMessagesLoaded = false;
+        vm.loadingData = false;
+
+        vm.timeGroup = {
+            amount: undefined,
+            unit: undefined,
+            buttonText: 'All Pending Retries',
+            selected: function () {
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: `selected` doesn't sound like an appropriate name. It implies a yes/no flag to me, but this returns a duration.

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 34]
```diff
@@ -0,0 +1,375 @@
...
+                return $moment.duration(vm.timeGroup.amount, vm.timeGroup.unit);;
+            }
+        };
+
+        notifier.subscribe($scope, function (event, data) {
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: It might be better to rename `data` to something meaningful here. Maybe `updatedTotalMessages`? Or something that reflects what it is. Makes the code easier to follow.

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 35]
```diff
@@ -0,0 +1,375 @@
...
+
+        notifier.subscribe($scope, function (event, response) {
+            vm.redirects = response.data;
+            refreshRedirects();
+        }, 'RedirectsUpdated');
+
+        var setSortButtonText = function (sort, direction) {
+            vm.sortButtonText = (sort === 'message_type' ? "Message Type" : "Time of Failure");
+            vm.sortDirection = direction;
+        }
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: Missing semicolon

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 36]
```diff
@@ -0,0 +1,375 @@
...
+
+        function removeResolvedMessages() {
+            vm.pendingRetryMessages = vm.pendingRetryMessages.filter(function (item) {
+                return !item.resolved;
+            });
+        }
+
+        function refreshRedirects() {
+            vm.pendingRetryMessages.map(function(obj) {
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: I don't think this is doing anything. `map` returns the new list, but we don't save the return result anywhere.

If we're simply iterating over the list to modify it and the result doesn't matter, I feel it would be better to use `forEach` to better represent that intention.

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 37]
```diff
@@ -0,0 +1,375 @@
...
+            vm.loadMoreResults();
+            vm.endpoints = endpointsService.getQueueNames();
+            vm.redirects = redirectService.getRedirects().data;
+        }
+
+        vm.noStatusPresent = function(message) {
+            return (!message.retried || !angular.isDefined(message.retried)) &&
+            (!message.resolved || !angular.isDefined(message.resolved)) && message.number_of_processing_attempts === 1;
+
+        }
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: Missing semicolon. Also there is an extra blank line.

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 38]
```diff
@@ -0,0 +1,375 @@
...
+
+            if (row.selected) {
+                vm.selectedIds.push(row.id);
+            } else {
+                vm.selectedIds.splice(vm.selectedIds.indexOf(row.id), 1);
+            }
+        };
+
+        vm.retrySelected = function () {
+            pendingRetryService.retryPendingRetriedMessages(vm.selectedIds).then(function() {
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: Should we have a `busy` status here? The `retryPendingRetriedMessages` method could take a while to execute, and the viewmodel properties on the UI are only updated in the success path. Is there a UI indication that the messages are busy being processed?

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 39]
```diff
@@ -0,0 +1,375 @@
...
+                });
+                toastService.showInfo('Selected messages were submitted for retry.');
+            }, function() {
+                toastService.showError('Failed to retry selected messages');
+            });
+        };
+
+        vm.retryAll = function () {
+            pendingRetryService.retryAllMessages(vm.filter.searchPhrase.physical_address, vm.filter.start, vm.filter.end).then(function() {
+                vm.selectedIds = [];
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: Same question here

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 40]
```diff
@@ -0,0 +1,375 @@
...
+                    item.retried = true;
+                });
+                toastService.showInfo('All filtered messages were submitted for retry.');
+            }, function() {
+                toastService.showError('Failed to retry all filtered messages');
+            });
+        };
+
+        vm.markAsResolvedAll = function () {
+            pendingRetryService.markAsResolvedAllMessages(vm.filter.searchPhrase ? vm.filter.searchPhrase.physical_address : '', vm.filter.start, vm.filter.end).then(function() {
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: And again

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js [hunk 41]
```diff
@@ -0,0 +1,375 @@
...
+        vm.retryAllConfirmationMessage = function () {
+            if (!vm.isQueueFilterEmpty()) {
+                return "Are you sure you want to retry " + vm.filteredTotal + " out of " + vm.total + " previously retried messages? If the selected messages were processed in the meanwhile, then duplicate messages will be produced.";
+            }
+
+            return "Bulk retry of messages can only be done for one queue at the time to avoid producing unwanted message duplicates.";
+        };
+
+        vm.markAsResolvedSelected = function () {
+            pendingRetryService.markAsResolvedMessages(vm.selectedIds).then(function() {
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: Same here

File: src/ServicePulse.Host/app/js/views/redirect/service.js [hunk 42]
```diff
@@ -0,0 +1,105 @@
...
+                    .success(function() {
+                        notifications.pushForCurrentRoute(success, 'info');
+                        getData();
+                    })
+                    .error(function () {
+                        notifications.pushForCurrentRoute(error, 'danger');
+                });
+            },
+            getTotalRedirects: function() {
+                return redirects.total;
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/redirect/service.js]: It confused me at first that these methods returned the data instead of a promise. Maybe return a promise that returns the data instead to be consistent?

`return $q.when(redirects.total);`

File: src/ServicePulse.Host/app/js/views/redirect/service.js [hunk 43]
```diff
@@ -0,0 +1,105 @@
...
+        function getData() {
+            var url = uri.join(scConfig.service_control_url, 'redirects');
+            return $http.get(url).then(function (response) {
+                notifier.notify('RedirectsUpdated', { total: response.headers('Total-Count'), data: response.data });
+                notifier.notify('RedirectMessageCountUpdated', response.headers('Total-Count'));
+            });
+        }
+
+        var redirectsUpdatedTimer = $interval(function () {
+            getData();
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/redirect/service.js]: Is this really something we have to query? Why can't we use SignalR to tell us when it changes?

File: src/ServicePulse.Host/app/js/views/redirect/service.js [hunk 44]
```diff
@@ -0,0 +1,105 @@
...
+        }
+
+        getData();
+
+        return {
+            createRedirect: function (sourceEndpoint, targetEndpoint, success, error) {
+                var url = uri.join(scConfig.service_control_url, 'redirects');
+                var promise = sendPromise(url, 'POST', { "fromphysicaladdress": sourceEndpoint, "tophysicaladdress": targetEndpoint }, success, error);
+                promise.then(function() {
+                    getData();
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/redirect/service.js]: `getData();` returns a promise, so this call could be replaced with:

```Javascript
return promise.then(getData);
```

Which then keeps the promise chain intact for future use or stacktraces.

File: src/ServicePulse.Host/app/js/views/redirect/service.js [hunk 45]
```diff
@@ -0,0 +1,105 @@
...
+                });
+
+                return promise;
+            },
+            updateRedirect: function (redirectId, sourceEndpoint, targetEndpoint, success, error) {
+                var url = uri.join(scConfig.service_control_url, 'redirects', redirectId);
+                var promise = sendPromise(url, 'PUT', { "id": redirectId, "fromphysicaladdress": sourceEndpoint, "tophysicaladdress": targetEndpoint }, success, error);
+                promise.then(function () {
+                    getData();
```
Comment (WilliamBZA) [inline src/ServicePulse.Host/app/js/views/redirect/service.js]: Same promise chaining option here

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js (hunk 31 above)
Comment (WojcikMike) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: It is indented 1 time from the level of `function controller`. Default indent in VS.

File: src/ServicePulse.Host/app/js/views/pending_retries/controller.js (hunk 36 above)
Comment (WojcikMike) [inline src/ServicePulse.Host/app/js/views/pending_retries/controller.js]: Yes it is effectively foreach

**Comment by WojcikMike** [comment]
@WilliamBZA I addressed your comments. With regards to SignalR I completely forgot about it - lame. I fixed that.

**Comment by WilliamBZA** [comment]
:+1: LGTM!

**Comment by WilliamBZA** [comment]
What is the deploy plan for this? If this is merged before the Control changes, will it cause issues?

**Comment by WojcikMike** [comment]
We should first merge SC changes and deploy them, then make sure that minimum SC version is set correctly. 

There will be also functional review done by @indualagarsamy @ramonsmits and possibly few others, so we might get some feedback related to functionalities or UI that can result in some code changes

**Comment by WojcikMike** [comment]
@WilliamBZA could you please merge?
